Divide by zero when sharding targets

### What's wrong?

One of our agents panicked with a divide by zero:

```
panic: runtime error: integer divide by zero

goroutine 503 [running]:
github.com/grafana/agent/component/discovery.(*DistributedTargets).Get(0xc002a9dd60)
	/src/agent/component/discovery/discovery.go:46 +0x29a
github.com/grafana/agent/component/prometheus/scrape.(*Component).distTargets(0xc003384d80, {0xcf7e320?, 0x0?, 0x0?}, {0xc002997781, 0x1a}, 0x0?)
	/src/agent/component/prometheus/scrape/scrape.go:316 +0xcd
github.com/grafana/agent/component/prometheus/scrape.(*Component).Run(0xc003384d80, {0x88179a8, 0xc002bef450})
	/src/agent/component/prometheus/scrape/scrape.go:220 +0x25c
github.com/grafana/agent/pkg/flow/internal/controller.(*ComponentNode).Run(0xc002b7f440, {0x88179a8, 0xc002bef450})
	/src/agent/pkg/flow/internal/controller/node_component.go:320 +0x111
github.com/grafana/agent/pkg/flow/internal/controller.newTask.func1()
	/src/agent/pkg/flow/internal/controller/scheduler.go:140 +0x6e
created by github.com/grafana/agent/pkg/flow/internal/controller.newTask in goroutine 234
	/src/agent/pkg/flow/internal/controller/scheduler.go:137 +0x138
```

This panic occurred less than 1 second after process start time.

### Steps to reproduce

:shrug: 

### System information

Linux AMD64

### Software version

Grafana Agent v0.38.1

### Configuration

_No response_

### Logs

```text
ts=2024-02-06T00:00:58.591802286Z level=info msg="finished node evaluation" controller_id="" trace_id=6d5b917cf460d9b2968892cb89f2f53f node_id=prometheus.scrape.conntrack_exporter duration=178.273µs
ts=2024-02-06T00:00:58.594385703Z level=info msg="Using pod service account via in-cluster config" component=discovery.kubernetes.kube_system_node_local_dns
ts=2024-02-06T00:00:58.594706096Z level=info msg="finished node evaluation" controller_id="" trace_id=6d5b917cf460d9b2968892cb89f2f53f node_id=discovery.kubernetes.kube_system_node_local_dns duration=2.884623ms
ts=2024-02-06T00:00:58.595620066Z level=info msg="finished node evaluation" controller_id="" trace_id=6d5b917cf460d9b2968892cb89f2f53f node_id=discovery.relabel.kube_system_node_local_dns duration=868.847µs
ts=2024-02-06T00:00:58.596143779Z level=info msg="finished node evaluation" controller_id="" trace_id=6d5b917cf460d9b2968892cb89f2f53f node_id=prometheus.scrape.kube_system_node_local_dns duration=495.838µs
ts=2024-02-06T00:00:58.596884909Z level=info msg="Using pod service account via in-cluster config" component=discovery.kubernetes.ip_space_usage_exporter
ts=2024-02-06T00:00:58.59718142Z level=info msg="finished node evaluation" controller_id="" trace_id=6d5b917cf460d9b2968892cb89f2f53f node_id=discovery.kubernetes.ip_space_usage_exporter duration=1.002161ms
ts=2024-02-06T00:00:58.598816391Z level=info msg="finished node evaluation" controller_id="" trace_id=6d5b917cf460d9b2968892cb89f2f53f node_id=discovery.relabel.ip_space_usage_exporter duration=1.513836ms
ts=2024-02-06T00:00:58.600214149Z level=info msg="finished node evaluation" controller_id="" trace_id=6d5b917cf460d9b2968892cb89f2f53f node_id=prometheus.scrape.ip_space_usage_exporter duration=314.507µs
ts=2024-02-06T00:00:58.60034833Z level=info msg="finished node evaluation" controller_id="" trace_id=6d5b917cf460d9b2968892cb89f2f53f node_id=otel duration=108.021µs
ts=2024-02-06T00:00:58.600398765Z level=info msg="finished complete graph evaluation" controller_id="" trace_id=6d5b917cf460d9b2968892cb89f2f53f duration=58.109039ms
ts=2024-02-06T00:00:58.600474871Z level=info msg="scheduling loaded components and services"
panic: runtime error: integer divide by zero

goroutine 503 [running]:
github.com/grafana/agent/component/discovery.(*DistributedTargets).Get(0xc002a9dd60)
	/src/agent/component/discovery/discovery.go:46 +0x29a
github.com/grafana/agent/component/prometheus/scrape.(*Component).distTargets(0xc003384d80, {0xcf7e320?, 0x0?, 0x0?}, {0xc002997781, 0x1a}, 0x0?)
	/src/agent/component/prometheus/scrape/scrape.go:316 +0xcd
github.com/grafana/agent/component/prometheus/scrape.(*Component).Run(0xc003384d80, {0x88179a8, 0xc002bef450})
	/src/agent/component/prometheus/scrape/scrape.go:220 +0x25c
github.com/grafana/agent/pkg/flow/internal/controller.(*ComponentNode).Run(0xc002b7f440, {0x88179a8, 0xc002bef450})
	/src/agent/pkg/flow/internal/controller/node_component.go:320 +0x111
github.com/grafana/agent/pkg/flow/internal/controller.newTask.func1()
	/src/agent/pkg/flow/internal/controller/scheduler.go:140 +0x6e
created by github.com/grafana/agent/pkg/flow/internal/controller.newTask in goroutine 234
	/src/agent/pkg/flow/internal/controller/scheduler.go:137 +0x138
```

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -69,6 +69,8 @@ Main (unreleased)
 - Fix an issue with static integrations-next marshaling where non singletons
   would cause `/-/config` to fail to marshal. (@erikbaranowski)
 
+- Fix divide-by-zero issue when sharding targets. (@hainenber) 
+
 ### Other changes
 
 - Removed support for Windows 2012 in line with Microsoft end of life. (@mattdurham)

--- a/component/discovery/discovery.go
+++ b/component/discovery/discovery.go
@@ -59,7 +59,7 @@ func (t *DistributedTargets) Get() []Target {
 			// back to owning the target ourselves.
 			res = append(res, tgt)
 		}
-		if len(peers) >= 1 && peers[0].Self {
+		if len(peers) == 0 || peers[0].Self {
 			res = append(res, tgt)
 		}
 	}

--- a/component/discovery/discovery.go
+++ b/component/discovery/discovery.go
@@ -43,7 +43,13 @@ func (t *DistributedTargets) Get() []Target {
 		return t.targets
 	}
 
-	res := make([]Target, 0, (len(t.targets)+1)/len(t.cluster.Peers()))
+	peerCount := len(t.cluster.Peers())
+	resCap := (len(t.targets) + 1) / peerCount
+	if peerCount == 0 {
+		resCap = len(t.targets) + 1
+	}
+
+	res := make([]Target, 0, resCap)
 
 	for _, tgt := range t.targets {
 		peers, err := t.cluster.Lookup(shard.StringKey(tgt.NonMetaLabels().String()), 1, shard.OpReadWrite)
@@ -53,7 +59,7 @@ func (t *DistributedTargets) Get() []Target {
 			// back to owning the target ourselves.
 			res = append(res, tgt)
 		}
-		if peers[0].Self {
+		if len(peers) >= 1 && peers[0].Self {
 			res = append(res, tgt)
 		}
 	}

--- a/component/discovery/discovery.go
+++ b/component/discovery/discovery.go
@@ -44,9 +44,9 @@ func (t *DistributedTargets) Get() []Target {
 	}
 
 	peerCount := len(t.cluster.Peers())
-	resCap := (len(t.targets) + 1) / peerCount
-	if peerCount == 0 {
-		resCap = len(t.targets) + 1
+	resCap := (len(t.targets) + 1)
+	if peerCount != 0 {
+		resCap = (len(t.targets) + 1) / peerCount
 	}
 
 	res := make([]Target, 0, resCap)

--- a/component/loki/source/podlogs/reconciler.go
+++ b/component/loki/source/podlogs/reconciler.go
@@ -131,7 +131,13 @@ func distributeTargets(c cluster.Cluster, targets []*kubetail.Target) []*kubetai
 		return targets
 	}
 
-	res := make([]*kubetail.Target, 0, (len(targets)+1)/len(c.Peers()))
+	peerCount := len(c.Peers())
+	resCap := (len(targets) + 1) / peerCount
+	if peerCount == 0 {
+		resCap = len(targets) + 1
+	}
+
+	res := make([]*kubetail.Target, 0, resCap)
 
 	for _, target := range targets {
 		peers, err := c.Lookup(shard.StringKey(target.Labels().String()), 1, shard.OpReadWrite)
@@ -141,7 +147,7 @@ func distributeTargets(c cluster.Cluster, targets []*kubetail.Target) []*kubetai
 			// back to owning the target ourselves.
 			res = append(res, target)
 		}
-		if peers[0].Self {
+		if len(peers) >= 1 && peers[0].Self {
 			res = append(res, target)
 		}
 	}

--- a/component/loki/source/podlogs/reconciler.go
+++ b/component/loki/source/podlogs/reconciler.go
@@ -147,7 +147,7 @@ func distributeTargets(c cluster.Cluster, targets []*kubetail.Target) []*kubetai
 			// back to owning the target ourselves.
 			res = append(res, target)
 		}
-		if len(peers) >= 1 && peers[0].Self {
+		if len(peers) == 0 || peers[0].Self {
 			res = append(res, target)
 		}
 	}

--- a/component/loki/source/podlogs/reconciler.go
+++ b/component/loki/source/podlogs/reconciler.go
@@ -132,9 +132,9 @@ func distributeTargets(c cluster.Cluster, targets []*kubetail.Target) []*kubetai
 	}
 
 	peerCount := len(c.Peers())
-	resCap := (len(targets) + 1) / peerCount
-	if peerCount == 0 {
-		resCap = len(targets) + 1
+	resCap := len(targets) + 1
+	if peerCount != 0 {
+		resCap = (len(targets) + 1) / peerCount
 	}
 
 	res := make([]*kubetail.Target, 0, resCap)

